hooks/hookexecution/entrypoint.go

```diff
@@ -0,0 +1,131 @@
+package hookexecution
+
+import (
+	"context"
+	"fmt"
+	"github.com/prebid/prebid-server/hooks"
+	"github.com/prebid/prebid-server/hooks/hookstage"
+	"github.com/prebid/prebid-server/hooks/invocation"
+	"net/http"
+	"sync"
+	"time"
+)
+
+type TimeoutError struct{}
+
+func (e TimeoutError) Error() string {
+	return fmt.Sprint("Hook execution timeout")
+}
+
+type RejectError struct {
+	Code   int
+	Reason string // is it needed or code is enough?
+}
+
+func (e RejectError) Error() string {
+	return fmt.Sprintf("Module rejected stage, reason: `%s`", e.Reason)
+}
+
+func ExecuteEntrypointStage(
+	invocationCtx *invocation.InvocationContext,
+	plan hooks.Plan[hookstage.Entrypoint],
+	req *http.Request,
+	body []byte,
+) (invocation.StageResult[hookstage.EntrypointPayload], []byte, error) {
+	var wg sync.WaitGroup
+	var stageResult invocation.StageResult[hookstage.EntrypointPayload]
+
+	payload := hookstage.EntrypointPayload{Request: req, Body: body}
+	done := make(chan struct{})
+
+	for _, groups := range plan {
+		groupResults := make([]invocation.HookResult[hookstage.EntrypointPayload], 0)
+		resp := make(chan invocation.HookResponse[hookstage.EntrypointPayload])
+
+		for _, moduleHook := range groups.Hooks {
+			moduleCtx := invocationCtx.ModuleContextFor(moduleHook.Module)
+			wg.Add(1)
+			go func(hw hooks.HookWrapper[hookstage.Entrypoint], moduleCtx *invocation.ModuleContext) {
+				defer wg.Done()
+				hookRespCh := make(chan invocation.HookResponse[hookstage.EntrypointPayload], 1)
+
+				select {
+				case <-done:
+					return
+				default:
+
+					go asyncHookCall(moduleCtx, hw.Hook, payload, groups.Timeout, invocationCtx.DebugEnabled, hookRespCh)
+					select {
+					case res := <-hookRespCh:
+						res.Result.ModuleCode = hw.Module
+						resp <- res
+					case <-time.After(groups.Timeout):
+						resp <- invocation.HookResponse[hookstage.EntrypointPayload]{
+							Result: invocation.HookResult[hookstage.EntrypointPayload]{ModuleCode: hw.Module},
+							Err:    TimeoutError{},
+						}
+					}
+				}
+			}(moduleHook, moduleCtx)
+		}
+
+		go func() {
+			wg.Wait()
+			close(resp)
+		}()
+
+		for r := range resp {
+			if r.Result.Reject {
+				close(done)
+				// todo: send metric
+				return stageResult, body, RejectError{Code: r.Result.NbrCode, Reason: r.Result.Message}
+			}
+
+			if r.Err != nil {
+				// todo: process error?
+				// todo: send metric
```

Comment:
In case of error, as currently written the result will still drop down into `groupResults` and processed for ChangeSets. Is this desired, or should we put the append below into an else clause to exclude errors? The "todo: process error" will likely result in a better solution in any case.

Reply:
We drop everything into `groupResults` because we need all kind of results for final stage outcomes. That information will be used for debug output later. In case a module returns an error we also expect that it would return an empty/nil ChangeSet as well or if the module decides to return some ChangeSet together with the error we do not prevent such operation. So we leave the decision to the module what should be returned